doc.rkt

```diff
@@ -0,0 +1,274 @@
+         "path-util.rkt"
+         "doc-trace.rkt"
+         racket/match
+         racket/class
+         racket/set
+         racket/list
+         racket/string
+         data/interval-map
+         syntax-color/module-lexer
+         syntax-color/racket-lexer
+         json)
+
+(struct Doc
+  (text
+   trace
+   uri
+   during-batch-change?
+   checked?)
+  #:transparent #:mutable)
+
+(define (send-diagnostics doc diag-lst)
+  (display-message/flush (diagnostics-message (Doc-uri doc) diag-lst)))
+
+;; the only place where really run check-syntax
+(define (doc-run-check-syntax doc)
+  (define (task)
+    (set-Doc-checked?! doc #f)
+    (match-define (list new-trace diags)
+      (check-syntax (uri->path (Doc-uri doc)) (Doc-text doc) (Doc-trace doc)))
+    (send-diagnostics doc diags)
+    (set-Doc-trace! doc new-trace)
+
+    (set-Doc-checked?! doc #t))
+
+  (scheduler-push-task! (Doc-uri doc) task))
+
+(define (lazy-check-syntax doc)
+  (when (not (Doc-during-batch-change? doc))
+    (doc-run-check-syntax doc)))
+
+(define (doc-checked? doc)
+  (Doc-checked? doc))
+
+(define (new-doc uri text)
+  (define doc-text (new lsp-editor%))
+  (send doc-text insert text 0)
+  ;; the init trace should not be #f
+  (define doc-trace (new build-trace% [src (uri->path uri)] [doc-text doc-text] [indenter #f]))
+  (define doc (Doc doc-text doc-trace uri #f #f))
+  (lazy-check-syntax doc)
+  doc)
+
+(define (doc-reset! doc new-text)
+  (define doc-text (Doc-text doc))
+  (define doc-trace (Doc-trace doc))
+
+  (send doc-text erase)
+  (send doc-trace reset)
+  (send doc-text insert new-text 0)
+  (lazy-check-syntax doc))
+
+(define (doc-update! doc st-ln st-ch ed-ln ed-ch text)
+  (define doc-text (Doc-text doc))
+  (define doc-trace (Doc-trace doc))
+
+  (define st-pos (doc-pos doc st-ln st-ch))
+  (define end-pos (doc-pos doc ed-ln ed-ch))
+  (define old-len (- end-pos st-pos))
+  (define new-len (string-length text))
+
+  ;; try reuse old information as the check-syntax can fail
+  ;; and return the old build-trace% object
+  (cond [(> new-len old-len) (send doc-trace expand end-pos (+ st-pos new-len))]
+        [(< new-len old-len) (send doc-trace contract (+ st-pos new-len) end-pos)])
+  (send doc-text replace text st-pos end-pos)
+  (lazy-check-syntax doc))
+
+(define-syntax-rule (doc-batch-change doc expr ...)
+  (let ()
+    (set-Doc-during-batch-change?! doc #t)
+    expr ...
+    (set-Doc-during-batch-change?! doc #f)
+    (lazy-check-syntax doc)))
+
+(define (doc-pos doc line ch)
+  (send (Doc-text doc) line/char->pos line ch))
+
+(define (doc-line/ch doc pos)
+  (match-define (list line char) (send (Doc-text doc) pos->line/char pos))
+  (values line char))
+
+(define (doc-line-start-pos doc line)
+  (send (Doc-text doc) line-start-pos line))
+
+(define (doc-line-end-pos doc line)
+  (send (Doc-text doc) line-end-pos line))
+
+(define (doc-endpos doc)
+  (send (Doc-text doc) end-pos))
+
+(define (doc-find-containing-paren doc pos)
```

Comment:
What's the purpose of this function? Checking paren balancing and stop invalid file?

Reply:
It's a new wrapper for the existing `find-containing-paren`. The function finds the closest pair of brackets the containing `pos` If I understand it correctly.